aten/src/ATen/native/sparse/SparseMatMul.cpp

```diff
@@ -0,0 +1,231 @@
+    const TensorAccessor<int64_t, N>& indices,
+    int64_t dim) {
+  auto nnz = indices.size(0);
+  std::vector<int64_t> csr(dim + 1, int64_t(0));
+  int64_t last_i = 0;
+  for (int64_t index = 0; index < nnz; index++) {
+    int64_t i = indices[index];
+    if (i == last_i) {
+      csr[last_i + 1] += 1;
+    } else {
+      for (int64_t j = last_i; j < i + 1; j++) {
+        csr[j + 1] = csr[last_i + 1];
+      }
+      last_i = i;
+      csr[last_i + 1] += 1;
+    }
+  }
+  for (int64_t j = last_i; j < dim; j++) {
+    csr[j + 1] = csr[last_i + 1];
+  }
+  return csr;
+}
+
+template <typename scalar_t>
+void sparse_matmul_kernel(
+    Tensor& result,
+    const Tensor& self,
+    const Tensor& other) {
+  auto indices_a = self._indices().contiguous();
+  auto values_a = self._values().contiguous();
+
+  auto indices_b = other._indices().contiguous();
+  auto values_b = other._values().contiguous();
+
+  auto nnz_a = values_a.size(0);
+  auto nnz_b = values_b.size(0);
+  auto indices_accessor_a = indices_a.accessor<int64_t, 2>();
+  auto indices_accessor_b = indices_b.accessor<int64_t, 2>();
+
+  auto values_accessor_a = values_a.accessor<scalar_t, 1>();
+  auto values_accessor_b = values_b.accessor<scalar_t, 1>();
+
+  auto result_indices = result._indices().contiguous();
+  auto result_values = result._values().contiguous();
+
+  std::map<std::pair<int64_t, int64_t>, scalar_t> d;
+  if (self.is_coalesced() && other.is_coalesced()) {
+    for (int64_t n1 = 0; n1 < nnz_a; n1++) {
+      auto r2 = indices2csr(indices_accessor_b[0], other.size(0));
+      for (int n2 = r2[indices_accessor_a[1][n1]];
+           n2 < r2[indices_accessor_a[1][n1] + 1];
+           n2++) {
+        auto current_index = std::make_pair(
+            indices_accessor_a[0][n1], indices_accessor_b[1][n2]);
+        d[current_index] += values_accessor_a[n1] * values_accessor_b[n2];
+      }
+    }
+  } else {
+    for (int64_t n1 = 0; n1 < nnz_a; n1++) {
+      for (int n2 = 0; n2 < nnz_b; n2++) {
+        if (indices_accessor_b[0][n2] == indices_accessor_a[1][n1]) {
+          auto current_index = std::make_pair(
+              indices_accessor_a[0][n1], indices_accessor_b[1][n2]);
+          d[current_index] += values_accessor_a[n1] * values_accessor_b[n2];
+        }
+      }
+    }
+  }
+  int64_t nnz_result = d.size();
+  std::vector<int64_t> values_size = {nnz_result};
+  result_indices.resize_({2, nnz_result});
+  result_values.resize_(values_size);
+
+  auto result_indices_accessor = result_indices.accessor<int64_t, 2>();
+  auto result_values_accessor = result_values.accessor<scalar_t, 1>();
+
+  int64_t index = 0;
+  for (auto kv : d) {
+    auto idx = kv.first;
+    result_indices_accessor[0][index] = idx.first;
+    result_indices_accessor[1][index] = idx.second;
+    result_values[index] = kv.second;
+    index++;
+  } 
+}
+
+template <typename scalar_t, bool grad_by_row>
+void sparse_matmul_kernel_grad(Tensor& result, const Tensor& x) {
```

Comment:
Dumb question: why do you need a separate kernel for grad? Naively, I would have expected the backwards of matmul to also be matmul, and so you ought to be able to reuse the regular sparse matmul kernel here.

Reply:
That is correct. @aocsa , if your forward is `C = AB`, then `A_grad = C_grad B^T, B_grad = A^T C_grad`.

Reply:
Yes, you are right,  problem is that there is not the concept of `fill_value` in the current COO tensors.  In some cases, when you get zeros in the  output  `C = AB` and they are not represented,  the `C_grad` tensor which is a sparse tensor has less values than it should be. That's why I did a custom implementation to get both `A_grad = C_grad  B^T, B_grad = A^T C_grad`

Reply:
But is it actually bad to reduce a number of elements to backprop through? A hypothetical example is when all the rows of `A` are orthogonal to all the columns of `B`, so then `AB = C = 0`.

Reply:
Hmm, I'm not exactly following how `fill_value` would solve the problem. I agree there is a  problem though!

Imagine that we do a sparse-sparse vector dot product l = v1 v2. We get a single scalar output, and when we do backwards, the gradient, in the dense case, would propagate to all entries of the input vectors. That is to say, v1_grad and v2_grad are dense, even if v1 and v2 are sparse! Typically, this isn't what people want, because it's easier to understand the efficiency properties of sparse operations if their backwards have the same sparsity properties. So we just ignore the contributions of the zero entries: it is more of a "mask out zero entries and then dot product" operation than the true mathematical operation.

I was hoping to see some discussion of this in the RFC, but I didn't see it anywhere (however, I didn't look particularly hard).

Reply:
> easier to understand the efficiency properties of sparse operations if their backwards have the same sparsity properties

That is a good point / design principle that should be made explicitly somewhere.

"default fill value is zero" makes sense as an interpretation for operations where the answer is typically the same as for a "masked tensor" kind of approach.

For operations that would change the sparsity pattern, it may depend on the task at hand if the "mask" or "zero" assumption should be made.

Reply:
One design rule @pearu just mentioned is:
- in the main `torch` namespace, sparse tensors should do the same as dense tensors
- in the `torch.sparse` namespace, sparse tensors can have different behaviour (e.g. "masked" sparsity-preserving matmul). that's also the case for `sparse.softmax`